In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df = pd.read_csv("diabetes.csv")

In [3]:
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

The data appears to have missing values. The missing values are mentioned as Zeros in columns 2 to 7.
For ease in calculations these zero values are converted to Nan.
Note : The column "Pregnancies" is allowed to have zero values because Zero pregnancies is a valid possible scenario.

In [4]:
columns = ['Glucose', 'BloodPressure', 'SkinThickness','Insulin', 'BMI','DiabetesPedigreeFunction', 'Age']
df[columns] = df[columns].replace(0,np.nan)

In [5]:
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
3              1     89.0           66.0           23.0     94.0  28.1   
4              0    137.0           40.0           35.0    168.0  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10    101.0           76.0           48.0    180.0  32.9   
764            2    122.0           70.0           27.0      NaN  36.8   
765            5    121.0           72.0           23.0    112.0  26.2   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

#### Correlation Matrix

In [6]:
df.corr()

Pregnancies   Glucose  BloodPressure  SkinThickness  \
Pregnancies                  1.000000  0.128135       0.214178       0.100239   
Glucose                      0.128135  1.000000       0.223192       0.228043   
BloodPressure                0.214178  0.223192       1.000000       0.226839   
SkinThickness                0.100239  0.228043       0.226839       1.000000   
Insulin                      0.082171  0.581186       0.098272       0.184888   
BMI                          0.021719  0.232771       0.289230       0.648214   
DiabetesPedigreeFunction    -0.033523  0.137246      -0.002805       0.115016   
Age                          0.544341  0.267136       0.330107       0.166816   
Outcome                      0.221898  0.494650       0.170589       0.259491   

                           Insulin       BMI  DiabetesPedigreeFunction  \
Pregnancies               0.082171  0.021719                 -0.033523   
Glucose                   0.581186  0.232771                  0.137246   
BloodPressure             0.098272  0.289230                 -0.002805   
SkinThickness             0.184888  0.648214                  0.115016   
Insulin                   1.000000  0.228050                  0.130395   
BMI                       0.228050  1.000000                  0.155382   
DiabetesPedigreeFunction  0.130395  0.155382                  1.000000   
Age                       0.220261  0.025841                  0.033561   
Outcome                   0.303454  0.313680                  0.173844   

                               Age   Outcome  
Pregnancies               0.544341  0.221898  
Glucose                   0.267136  0.494650  
BloodPressure             0.330107  0.170589  
SkinThickness             0.166816  0.259491  
Insulin                   0.220261  0.303454  
BMI                       0.025841  0.313680  
DiabetesPedigreeFunction  0.033561  0.173844  
Age                       1.000000  0.238356  
Outcome                   0.238356  1.000000

#### Data Preprocessing

In [7]:
df.isnull().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

The missing values shall be replaced with the corresponding means. 
Note: Since there are too many missing values(almost 50%) for "Insulin", using mean to replace these values is not advisable. Therefore, "Insulin" shall be excluded from further calculations.

In [8]:
df['Glucose'].fillna(value=df['Glucose'].mean(),inplace=True)
df['BloodPressure'].fillna(value=df['BloodPressure'].mean(),inplace=True)
df['BMI'].fillna(value=df['BMI'].mean(),inplace=True)
df['SkinThickness'].fillna(value=df['SkinThickness'].mean(),inplace=True)

In [9]:
df.isnull().sum()

Pregnancies                   0
Glucose                       0
BloodPressure                 0
SkinThickness                 0
Insulin                     374
BMI                           0
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

#### Selecting x and y variables

In [10]:
x= df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'BMI', 'DiabetesPedigreeFunction', 'Age']].values

In [11]:
x

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [12]:
y=df.iloc[:,-1].values

In [13]:
y

array([1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,

#### Test Train Splitting

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

#### Standardization of Data

In [16]:
from sklearn.preprocessing import  StandardScaler

In [17]:
sc = StandardScaler()

In [18]:
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test)

#### Applying the Model
We shall use Logistical Regression for this model, since the data output is a Classification Type data

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
classifier = LogisticRegression()

In [21]:
classifier.fit(x_train, y_train)

LogisticRegression()

In [22]:
y_pred = classifier.predict(x_test)
y_pred_train = classifier.predict(x_train)

#### Accuracy Check

In [23]:
from sklearn.metrics import confusion_matrix, accuracy_score

In [24]:
# Accuracy of Training Data
accuracy_score(y_train, y_pred_train)

0.7638436482084691

In [25]:
# Accuracy of Test Data
accuracy_score(y_test, y_pred)

0.8116883116883117

In [26]:
#Confusion Matrix
confusion_matrix(y_test,y_pred)

array([[97, 10],
       [19, 28]], dtype=int64)

### Conclusion 

A prediction model based on Logistic Regression was used to predict the outcome of diabetes in patients. The data set was first filtered and the missing values were either eliminitaed or replaced with the mean. 
The correlation matrix revealed some extent of correlation bwtween the outcome and all the variables. Thus it was decided to include all the columns for analysis. However, the "Insulin" column had almost 50% of the values missing. So, it was excluded from calculations.
Accuracy of Train Data = 76%
Accuracy of Test Data  = 81%